feat(bazel): support ts_library targets as entry-points for ng_package

devversion opened the pull request:
Within an Angular package, it can happen that there are
entry-points which do not contain features that belong into
an `@NgModule` or need metadata files to be generated.

For example: the `cdk`, `cdk/testing` and `cdk/coercion`
entry-points. Besides other entry-points in the `cdk`
package, those entry-points do not need metadata to
be generated and no not use the `ng_module` rule.

Currently the "ng_package" rule properly picks up such
entry-points and builds bundles, does downleveling etc.
The only thing it misses is that no `package.json` files
are generated for the entry-point. This means that consumers
will not be able to use these entry-points built with "ts_library"
(except accessing the individual bundlings explicitly).

The "ng_package" rule should follow the full APF specification
for such entry-points. Partially building bundles and doing the
downleveling is confusing and a breaking issue.

---

The motifivation of supporting this (besides making the
rule behavior consistent; the incomplete output is not
acceptable), is that using the "ng_module" rule does
not make sense to be used for non-Angular entry-points.

Especially since it depends on Angular packages to
be specified as Bazel action inputs just to compile
vanilla TypeScript with `@angular/compiler-cli`.

jelbourn commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -228,22 +228,49 @@ def _ng_package_impl(ctx):
+            typings_path = ng_module_metadata.typings_file.path
+            metadata_file = ng_module_metadata.metadata_file
         else:
+            output_dir = "/".join([
```
Add a comment about what kinds of rules your expecting for each of these branches?

devversion replied:
Done. I had to place the comments into the branches as otherwise buildifier formats it weirdly.

devversion committed:
feat(bazel): support ts_library targets as entry-points for ng_package

Within an Angular package, it can happen that there are
entry-points which do not contain features that belong into
an `@NgModule` or need metadata files to be generated.

For example: the `cdk`, `cdk/testing` and `cdk/coercion`
entry-points. Besides other entry-points in the `cdk`
package, those entry-points do not need metadata to
be generated and no not use the `ng_module` rule.

Currently the "ng_package" rule properly picks up such
entry-points and builds bundles, does downleveling etc.
The only thing it misses is that no `package.json` files
are generated for the entry-point. This means that consumers
will not be able to use these entry-points built with "ts_library"
(except accessing the individual bundlings explicitly).

The "ng_package" rule should follow the full APF specification
for such entry-points. Partially building bundles and doing the
downleveling is confusing and a breaking issue.

The motifivation of supporting this (besides making the
rule behavior consistent; the incomplete output is not
acceptable), is that using the "ng_module" rule does
not make sense to be used for non-Angular entry-points.

Especially since it depends on Angular packages to
be specified as Bazel action inputs just to compile
vanilla TypeScript with `@angular/compiler-cli`.

alexeagle reviewed (approved):
nice job as always

josephperrott reviewed (approved):
LGTM

devversion committed:
fixup! feat(bazel): support ts_library targets as entry-points for ng_package

Add debug messages

devversion commented on packages/bazel/src/ng_package/ng_package.bzl:
```diff
@@ -228,22 +228,60 @@ def _ng_package_impl(ctx):
+
             # fallback to a reasonable default
-            flat_module_out_file = "index.js"
+            index_file = "index.js"
```
Great idea. Done!
